upstairs/src/upstairs.rs

```diff
@@ -901,71 +880,53 @@ impl Upstairs {
     /// If this Upstairs is [UpstairsConfig::read_only], this function will move
     /// any Downstairs from
     /// `DsState::Connecting { state:  NegotiationState::LiveRepairReady, .. }`
     /// back to [DsState::Active] without actually performing any repair.
     pub(crate) fn on_repair_check(&mut self) {
-        info!(self.log, "Checking if live repair is needed");
         if !matches!(self.state, UpstairsState::Active) {
-            info!(self.log, "inactive, no live repair needed");
-            self.repair_check_deadline = None;
             return;
         }
 
         if self.cfg.read_only {
-            info!(self.log, "read-only, no live repair needed");
             // Repair can't happen on a read-only downstairs, so short circuit
             // here. There's no state drift to repair anyway, this read-only
             // Upstairs wouldn't have caused any modifications.
             for c in self.downstairs.clients.iter_mut() {
                 c.skip_live_repair(&self.state);
             }
-            self.repair_check_deadline = None;
             return;
         }
 
-        // Verify that all downstairs and the upstairs are in the proper state
-        // before we begin a live repair.
-        let repair_in_progress = self.downstairs.live_repair_in_progress();
+        // If we're already doing live-repair, then we can't start live-repair
+        if self.downstairs.live_repair_in_progress() {
+            return;
+        }
 
-        let any_in_repair_ready = self.downstairs.clients.iter().any(|c| {
+        // If no one is LiveRepairReady, then we can't start live-repair
+        if !self.downstairs.clients.iter().any(|c| {
             matches!(
                 c.state(),
                 DsState::Connecting {
                     state: NegotiationState::LiveRepairReady,
                     ..
                 }
             )
-        });
+        }) {
+            return;
+        }
 
-        if repair_in_progress {
-            info!(self.log, "Live Repair already running");
-            // Queue up a later check if we need it
-            if any_in_repair_ready {
-                self.repair_check_deadline =
-                    Some(Instant::now() + REPAIR_CHECK_INTERVAL);
-            } else {
-                self.repair_check_deadline = None;
-            }
-        } else if !any_in_repair_ready {
-            self.repair_check_deadline = None;
-            info!(self.log, "No Live Repair required at this time");
-        } else if !self.downstairs.start_live_repair(
+        // Try to start live-repair, logging if it fails
+        if !self.downstairs.start_live_repair(
             &self.state,
             self.ddef.get_def().unwrap().extent_count(),
         ) {
             // It's hard to hit this condition; we need a Downstairs to be in
             // LiveRepairReady, but for no other downstairs to be in Active.
             warn!(self.log, "Could not start live repair, trying again later");
```

Comment:
How fast will this check loop?
Could this warning (which I do like to see) end up spamming the log file?

Reply:
Yes, this will spam the logs if it fails.  Frequency depends on how often events are hitting `Upstairs::apply`; minimum of 1/sec for stat updates, but it could be more if the Upstairs is spamming IO.  I don't have strong opinions on the best way to handle this, any ideas?

My weak inclination is to remove all logging from the failure path; we'll see enough other information in the logs to triangulate what's going wrong in the rare case where we want to live-repair but there are no Active downstairs.

Reply:
I agree with your weak inclination here, at least for this message.  Any message spamming the log is almost always worse, and in this case there should be enough other information (seeing DS transitions) that we can determine what happened.

Reply:
Okay, I've removed the logging - this is now just normal invariant maintenance, so it doesn't need to be treated specially!